### PR TITLE
feat: canonicalize `self` in `BitPackedArray::take` if large number of indices

### PR DESCRIPTION
**Pull request by lwwmanning**

### CODE REVIEW

**Comment by lwwmanning** [comment]
possible alternative impl of #1033

File: vortex-array/src/compute/take.rs [hunk 1]
```diff
@@ -10,6 +11,25 @@ pub trait TakeFn {
...
+
+    if !indices.dtype().is_int() || indices.dtype().is_nullable() {
+        vortex_bail!(
+            "Take indices must be a non-nullable integer type, got {}",
+            indices.dtype()
+        );
+    }
+
+    // If the indices are large enough, it's faster to canonicalize the indices and take those.
+    if indices.len() >= array.len() && PrimitiveArray::try_from(indices).is_err() {
```
Comment (danking) [inline vortex-array/src/compute/take.rs]: Are there any encodings that don't immediately convert the indices into a PrimitiveArray already? It seems a bit confusing to have this at the top-level but then have every encoding `into_primitive` anyway.

**Comment by lwwmanning** [comment]
tldr: decompression time down by ~25% for taxi/Euro2016, down ~50% for Bimbo/CMS/Food/HashTags, roughly flat for AirlineSentiment/Arade

```
decompress time/taxi    300901075 ns (4422884.719375014)        415849693.65 ns (5489439.483125001)     0.72
decompress time/AirlineSentiment        39633.29355277468 ns (175.943787235934) 38369.79170532914 ns (64.94147826156404)        1.03
decompress time/Arade   670599878.1 ns (6737559.532500029)      710661378.4 ns (21263203.21749997)      0.94
decompress time/Bimbo   2643467417 ns (333733330.16499996)      6146293028.5 ns (388309402.5999999)     0.43
decompress time/CMSprovider     7782017255.3 ns (119895042.70125008)    14693000868.6 ns (167426700.9074993)    0.53
decompress time/Euro2016        324697100.4 ns (2487771.782499999)      434308196.45 ns (1051625.4831250012)    0.75
decompress time/Food    278950325.55 ns (2474860.9900000095)    583034271.3 ns (1555255.6825000644)     0.48
decompress time/HashTags        688457951.5 ns (2789456.4650000334)     1162135494.7 ns (3129561.2625000477)    0.59
decompress time/TPC-H l_comment chunked without fsst    37736745.149722226 ns (340300.80451736227)      50389607.982579365 ns (950075.2169419602)       0.75
decompress time/TPC-H l_comment chunked 91673955.8176984 ns (137902.40186210722)        88119541.55452381 ns (344583.9479404837)        1.04
decompress time/TPC-H l_comment canonical       93116398.72350529 ns (219771.85710962117)       87804234.6486012 ns (486837.8035375625) 1.06
```

**Comment by lwwmanning** [comment]
for the bitpacked_take benchmark (vs develop):

Screenshot 2024-10-14 at 20 19 28
Screenshot 2024-10-14 at 20 19 18

File: encodings/fastlanes/src/bitpacking/compute/take.rs [hunk 2]
```diff
@@ -4,16 +4,31 @@ use fastlanes::BitPacking;
...
+// assuming the buffer is already allocated (which will happen at most once) then unpacking
+// all 1024 elements takes ~8.8x as long as unpacking a single element on an M2 Macbook Air.
+// see https://github.com/spiraldb/vortex/pull/190#issue-2223752833
+const UNPACK_CHUNK_THRESHOLD: usize = 8;
+const PATCH_THRESHOLD: usize = 64;
+
 impl TakeFn for BitPackedArray {
     fn take(&self, indices: &Array) -> VortexResult<Array> {
+        // If the indices are large enough, it's faster to flatten and take the primitive array.
+        if indices.len() * UNPACK_CHUNK_THRESHOLD > self.len() {
```
Comment (danking) [inline encodings/fastlanes/src/bitpacking/compute/take.rs:24]: I don't think the UNPACK_CHUNK_THRESHOLD is necessarily the right factor. Suppose there are 8K indices but they're all <= 1024. These would get chunked together and only trigger decompression of the first chunk.

When the indices are "chunk-contiguous" (i.e. all indices in [0, 1024) come first, then [1024, 2048), etc.), `take_primitive` already makes the correct choice between decompressing a chunk and individuals.

However, regardless of order, if the indices are sufficiently large, manipulating them dominates runtime. There's basically a per-index cost to construct `relative_indices` and another per-index cost to copy the patches to the right place. It's something like

```
if (n_indices * (chunk_and_relativize_cost + patch_cost * sparsity))
   >
   (decompress_cost * n_values)
```

**Comment by danking** [comment]
decompress summary

| Benchmark suite | Current: e84a63cc7d9d6ef64695f07399a729a6bfe0b8e0 | Previous: 8d9f62064c63350bccf18be76b308f435433c0a1 | Ratio |
|-|-|-|-|
| `taxi` | `300901075` ns  | `415849693.65` ns  | `0.72` |
| `AirlineSentiment` | `39633.29355277468` ns  | `38369.79170532914` ns  | `1.03` |
| `Arade` | `670599878.1` ns  | `710661378.4` ns  | `0.94` |
| `Bimbo` | `2643467417` ns  | `6146293028.5` ns  | `0.43` |
| `CMSprovider` | `7782017255.3` ns  | `14693000868.6` ns  | `0.53` |
| `Euro2016` | `324697100.4` ns  | `434308196.45` ns  | `0.75` |
| `Food` | `278950325.55` ns  | `583034271.3` ns  | `0.48` |
| `HashTags` | `688457951.5` ns  | `1162135494.7` ns  | `0.59` |
| `TPC-H l_comment chunked` | `91673955.8176984` ns  | `88119541.55452381` ns  | `1.04` |

File: encodings/fastlanes/src/bitpacking/compute/take.rs (hunk 2 above)
Comment (lwwmanning) [reply encodings/fastlanes/src/bitpacking/compute/take.rs:24]: Yeah, I agree that this is a bit crude / could be further refined, but it's a decent approximation.

The reason that I picked this same threshold thing is that if the indices are distributed uniformly then this will be ~optimal (even if they're sorted; but definitely if they're not).

The worst case would be if they're all in [0, 1024)... will add a bench for that

**Comment by danking** [comment]
I think the DictArray change still dominates b/c it fixes this and also any other compression under dict.

**Comment by lwwmanning** [comment]
yeah, this is much faster. I think the grouping of indices is actually very slow, and there's more juice to squeeze there

Screenshot 2024-10-14 at 20 40 38

**Comment by lwwmanning** [comment]
> I think the DictArray change still dominates b/c it fixes this and also any other compression under dict.

I think both. This changes top-level `take` such that DictArray would canonicalize if `codes.len()` is greater than `values.len()`, which should generally be the case right now, but won't be with shared dictionary

**Comment by lwwmanning** [comment]
also filed https://github.com/spiraldb/vortex/issues/1039 as a FLUP

File: vortex-array/src/compute/take.rs (hunk 1 above)
Comment (robert3005) [reply vortex-array/src/compute/take.rs]: Right now most of the time you flatten the indices but I think we shoulnd't do that. There's a couple of optimizations here to explore and we should first look at them before changing behaviour of top level function.

**Comment by danking** [comment]
🤦 didn't realize benches were running and readied the label. It says cancelled but the original run, https://github.com/spiraldb/vortex/actions/runs/11347211294, is still running.

**Review by danking — APPROVED** [review]
